File: src/ai/flows/quality-assessment.ts
// This file uses server-side code.
'use server';

/**
 * @fileOverview Analyzes product photos for quality indicators and suggests improvements.
 *
 * - analyzeProductPhoto - A function that handles the analysis of product photos.
 * - AnalyzeProductPhotoInput - The input type for the analyzeProductPhoto function.
 * - AnalyzeProductPhotoOutput - The return type for the analyzeProductPhoto function.
 */

import {ai} from '@/ai/genkit';
import {z} from 'genkit';

const AnalyzeProductPhotoInputSchema = z.object({
  photoDataUri: z
    .string()
    .describe(
      "A photo of the product, as a data URI that must include a MIME type and use Base64 encoding. Expected format: 'data:<mimetype>;base64,<encoded_data>'."
    ),
  productDescription: z.string().describe('The description of the product.'),
});
export type AnalyzeProductPhotoInput = z.infer<typeof AnalyzeProductPhotoInputSchema>;

const AnalyzeProductPhotoOutputSchema = z.object({
  qualityScore: z
    .number()
    .describe('A score indicating the quality of the product photo (0-100).'),
  suggestedImprovements: z
    .string()
    .describe('Suggestions for improving the product photo.'),
});
export type AnalyzeProductPhotoOutput = z.infer<typeof AnalyzeProductPhotoOutputSchema>;

export async function analyzeProductPhoto(
  input: AnalyzeProductPhotoInput
): Promise<AnalyzeProductPhotoOutput> {
  return analyzeProductPhotoFlow(input);
}

const prompt = ai.definePrompt({
  name: 'analyzeProductPhotoPrompt',
  input: {schema: AnalyzeProductPhotoInputSchema},
  output: {schema: AnalyzeProductPhotoOutputSchema},
  prompt: `You are an expert in product photography and visual merchandising.

You will analyze the provided product photo and description to assess its quality and provide actionable suggestions for improvement.

Consider factors such as lighting, composition, focus, clarity, and overall appeal.

Product Description: {{{productDescription}}}
Product Photo: {{media url=photoDataUri}}

Based on your analysis, provide a quality score (0-100) and specific suggestions for improving the photo to better showcase the product and attract potential buyers.
`,
});

const analyzeProductPhotoFlow = ai.defineFlow(
  {
    name: 'analyzeProductPhotoFlow',
    inputSchema: AnalyzeProductPhotoInputSchema,
    outputSchema: AnalyzeProductPhotoOutputSchema,
  },
  async input => {
    const {output} = await prompt(input);
    return output!;
  }
);

File: backend/src/ai/flows/quality_assessment.py
# analyze_product_photo.py
"""
Analyzes product photos for quality indicators and suggests improvements.

- analyze_product_photo: Handles product photo analysis.
- AnalyzeProductPhotoInput: Input type.
- AnalyzeProductPhotoOutput: Output type.
"""
from backend.data_types_class import AnalyzeProductPhotoInput, AnalyzeProductPhotoOutput
import base64

# Mock AI analysis function (replace with real AI/vision integration)
async def ai_analyze_product_photo(
    input: AnalyzeProductPhotoInput,
) -> AnalyzeProductPhotoOutput:
    # Validate Data URI
    try:
        header, encoded = input.photoDataUri.split(",", 1)
        _ = base64.b64decode(encoded)  # just ensures it's valid base64
    except Exception as e:
        raise ValueError("Invalid photoDataUri format.") from e

    # Mock scoring logic
    quality_score = 75  # Pretend the AI model rated it
    improvements = (
        "Improve lighting to reduce shadows, ensure the background is clean, "
        "and enhance sharpness for better clarity."
    )

    return AnalyzeProductPhotoOutput(
        qualityScore=quality_score, suggestedImprovements=improvements
    )


# Flow equivalent
async def analyze_product_photo(
    input: AnalyzeProductPhotoInput,
) -> AnalyzeProductPhotoOutput:
    return await ai_analyze_product_photo(input)




